src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: b.clone(),
+            },
+            FunctionType::new(vec![NAT], twoway(NAT)),
+            make_opaque(
+                B,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&b),
+            ),
+        )?;
+
+        let bb10 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        let bb11 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        // Internal wiring for DFGs
+        hugr.connect(entry_in, 0, mkpred, 0)?;
+        hugr.connect(mkpred, 0, entry_out, 0)?;
+
+        // CFG Wiring
+        hugr.connect(entry, 0, bb0, 0)?;
+        hugr.connect(entry, 0, bb1, 0)?;
+        hugr.connect(bb1, 0, bb10, 0)?;
+        hugr.connect(bb1, 0, bb11, 0)?;
+
+        hugr.connect(bb0, 0, exit, 0)?;
+        hugr.connect(bb10, 0, exit, 0)?;
+        hugr.connect(bb11, 0, exit, 0)?;
+
+        hugr.infer_extensions()?;
+
+        Ok(())
+    }
+
+    #[test]
+    fn multi_entry() -> Result<(), Box<dyn Error>> {
+        let mut hugr = Hugr::new(NodeType::open_extensions(ops::CFG {
+            signature: FunctionType::new(type_row![NAT], type_row![NAT]), // maybe add extensions?
+        }));
+        let cfg = hugr.root();
+        let exit = hugr.add_node_with_parent(
+            cfg,
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+        let entry = {
+            let dfb = ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: ExtensionSet::new(),
+            };
+
+            let entry = hugr.add_node_before(exit, NodeType::open_extensions(dfb))?;
```

Comment:
Is this not `let (entry, entry_in, entry_out) = create_with_io(.....)` ? (If you do `entry` before `exit` maybe)

Reply:
Hm, I thought about this but opted against it because defining entry before exit like that is taking advantage of implementation details and using `add_node_before` is more "the right thing"

Reply:
Sorry, I am barking up the wrong tree here. This is nothing like `create_with_io` - that makes an Input and Output (dataflow) nodes, this makes an entry (BasicBlock::DFB) and exit (BasicBlock::Exit) node.

That said, given you do this a lot across three or four tests, maybe you could just have a helper `create_with_entry_exit`....